---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# TWASviz

<!-- badges: start -->
<!-- badges: end -->

  The goal of TWASviz is to allow users to perform Transcriptome Wide Association Studies (TWASs) in R using sparse group lasso, and to import their own TWAS results and effectively visualize the output. This tool will also provide functions that visualize the gene expression data, the overlap of sets of important gene transcripts from multiple cell types or tissues, and the Gene Ontology term enrichments.
   Currently, the only TWAS visualization related R packages make Manhattan plots and focus on single TWASs at a time. However, with the development of cell type specific expression-quantitative trait loci and the importance of understanding the contribution of multiple tissues to disease, TWASviz provides several heatmap-type visualizations that help users compare and efficiently present the results of multiple cell type or tissue specific TWASs at once.
   
## Background

  TWASs are a powerful method for finding gene-trait associations. Simply associating individuals’ specific single nucleotide polymorphisms (SNPs) with a complex trait will highlight a lot of SNPs located in non-coding regions (this is a Genome Wide Association Study). Since non-coding regions are not transcribed, changes in non-coding regions change neither the structure nor functionality of the genes' resulting proteins. In contrast, TWASs associate gene transcript levels with phenotypes. This approach more directly implicates specific genes that functionally impact phenotype. As well, to perform a TWAS, you only need RNA-seq from a reference panel, and genotype data and phenotype data for your sample of interest. No need to perform any RNA sequencing on your sample of interest.

![Figure 1. TWAS overview.](./inst/extdata/TWAS_general.png)

  TWASviz is built to be compatible with the TWAS software [PrediXcan](https://github.com/hakyimlab/MetaXcan/wiki/Individual-level-PrediXcan:-introduction,-tutorials-and-manual). PrediXcan is can do all 3 steps in the above Figure 1a. However, the association that PrediXcan does is just a naive simple linear regression of predicated gene expression against phenotype data, one gene at a line. With TWASviz, PrediXcan is meant to do steps 1. Train and 2. Predict, and TWASviz is meant to do step 3. Associate (via sparse group lasso) and a new step 4. Visualize. TWASviz supplies functions that are meant to take in PrediXcan predicted gene expression data, and even has a couple of functions that are meant to deal directly with PrediXcan TWAS association results. Since PrediXcan is a very beginner friendly TWAS software, TWASviz is meant to attract PrediXcan users, invite them to try a different TWAS framework, and to give them some functions to produce interesting TWAS summary plots.

## TWASviz Overview

![Figure 2. TWASviz workflow.](./inst/extdata/TWASviz_workflow.png)

R version: 4.4.1 (2024-06-14 ucrt)
Platform: x86_64-w64-mingw32/x64
Running under: Windows 11 x64 (build 26100)


To install the latest version of the package:
```
install.packages("devtools")
library("devtools")
devtools::install_github("mding010705/TWASviz", build_vignettes = TRUE)
library("TWASviz")
```

To run the shinyApp: 

```
Under construction
```

Overview:

```
ls("package:TWASviz")
data(package = "TWASviz")
browseVignettes("TWASviz")
```

Available Functions:
corr_heatmap

preprocess_expressions_pathways

sgl_TWAS

sgl2adj_df

correlation_overlap_heatmap

gene_enrichment

goenrich_heatmap

volcano_plot

predixcan2adj_df


Contributions:

Megan Ding- Package author (conceptualized package, wrote code, and wrote tests)

Others that I don't have time to cite...

corr_heatmap

preprocess_expressions_pathways

sgl_TWAS

sgl2adj_df

correlation_overlap_heatmap

gene_enrichment

goenrich_heatmap

volcano_plot

predixcan2adj_df

## References

Gamazon, E. R., Wheeler, H. E., Shah, K. P., Mozaffari, S. V., Aquino-Michaels, K., Carroll, R. J., Eyler, A. E., Denny, J. C., Nicolae, D. L., Cox, N. J., & Im, H. K. (2015). A gene-based association method for mapping traits using reference transcriptome data. Nature Genetics, 47(9), 1091–1098. https://doi.org/10.1038/ng.3367 

Simon, N., Friedman, J., Hastie, T., & Tibshirani, R. (2013). A sparse-group lasso. Journal of Computational and Graphical Statistics, 22(2), 231–245. https://doi.org/10.1080/10618600.2012.681250 

Yu G (2024). “Thirteen years of clusterProfiler.” The Innovation, 5(6), 100722. doi:10.1016/j.xinn.2024.100722, https://doi.org/10.1016/j.xinn.2024.100722.

Wang, N., Ye, Z., & Ma, T. (2024). Tips: A novel pathway-guided joint model for transcriptome-wide association studies. Briefings in Bioinformatics, 25(6). https://doi.org/10.1093/bib/bbae587 



## Acknowledgements

This package was developed as part of an assessment for 2025 BCB410H: Applied Bioinformatics course at the University of Toronto, Toronto, CANADA. 

TWASviz welcomes issues, enhancement requests, and other contributions. To submit an issue, use the GitHub issues.
