---
title: "Introduction to TWASviz"
author: "Megan Ding"
date: "2025-11-28"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Introduction to TWASviz}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
## Introduction
`TWASviz` is an R package designed to produce important data exploration plots and Transcriptome Wise Association Studies (TWASs) with sparse group lasso (Simon et al., 2013). TWASs use imputed gene expression data and test for association between each gene’s expression levels against phenotype data. This process allows us to identify transcripts that are significantly correlated to a phenotype of interest. TWASviz will allow users to perform TWASs using sparse group lasso in R (Wang et al., 2024), and import their own TWAS results, and effectively visualize the output. 

This package is designed to use perform sparse group lasso TWASs on individual imputed gene expression data from the external software `PrediXcan` (Gamazon et al., 2015), as well as to process the linear regression TWAS output from `PrediXcan`. For more information on `PrediXcan` and its individual level gene expression imputation and TWAS association tests visit [their GitHub wiki](https://github.com/hakyimlab/MetaXcan/wiki/Individual-level-PrediXcan:-introduction,-tutorials-and-manual).

To download **TWASviz**, use the following commands:

```r
if (!requireNamespace("devtools")){
   install.packages("devtools")
}
require("devtools")
devtools::install_github("mding010705/TWASviz",
                         build_vignettes = TRUE)
library(TWASviz)
```

To list all functions available in the package:
```r
# ls("package:TWASviz") Lists all the functions from imported packages as well (too many)

# to list the exported functions in TWASviz
getNamespaceExports("TWASviz")
```

To list all sample datasets available in the package:
```r
data(package = "TWASviz")
```


## Components

![TWASviz workflow.](../inst/extdata/TWASviz_workflow.png){width=100%, height=100%}

## Application 1

Let's look at an analysis that starts from `PrediXcan` TWAS association output.

### Loading PrediXcan Association Results

The package comes with sample prediXcan association results (`predixcan_twasX` for sample `PrediXcan` association output).

```{r, warning=FALSE, message=FALSE}
library(TWASviz)
head(TWASviz::predixcan_twas1)
head(TWASviz::predixcan_twas2)
head(TWASviz::predixcan_twas3)
```

### Formatting Multiple PrediXcan TWAS Results

`predixcan2adj_df` is a function that reads in multiple `PrediXcan` association files and formats them in an adjacency-matrix-like dataframe, thresholding on the gene associations' p-values. The *NA* in the `adj_df` output means that this gene/pathway is not associated with your phenotype of interest in this tissue, but it is associated in a different tissue. This is possibly because this gene did not have a predicted gene expression in this tissue or that this gene was dropped after p-value thresholding.


```{r, warning=FALSE, message=FALSE}
adj_df <- TWASviz::predixcan2adj_df(predixcan_assoc_filenames = paste0("../inst/extdata/predixcan_twas", 1:6, ".txt"), 
                           pvalue_thresh = 0.02)

adj_df[50:55, ]
```


### Understanding the Relationship between Multiple TWASs

The `correlation_overlap_heatmap` function provides a heatmap for pairwise complete correlations for effect sizes of multiple gene sets (from TWASs) in the lower triangle, and number of pairwise complete effect sizes for multiple gene sets in the upper triangle. The diagonal displays the total number of effect sizes for a single gene set.

```{r, fig.width=8, fig.height=8}
TWASviz::correlation_overlap_heatmap(adj_df, tissue_names = c("a", "b", "c",
                                                     "d", "e", "f"))

```

### Understanding Significance and Effect Size Relationship 

`volcano_plot` produces a volcano plot of the effect size z-score against the -log10(pvalue). It labels and colors significant genes. The horizontal line is the significance threshold, and the vertical lines are at the normal 2.5 and 97.5th quantile thresholds. Notice that with `PrediXcan` association outputs, the systematic relationship between effect size and p-value is very noticeable due to the constant effect size.

```{r, fig.width=8, fig.height=10}
TWASviz::volcano_plot(betas_p_vals = TWASviz::predixcan_twas1[1:1000, c("zscore", "pvalue", "gene")], p_thresh = 0.0006)
```

### Gene Enrichment

Having a set of significant genes is not useful when you can't summarize their functional associations. `gene_enrichment` allows you to use `clusterProfiler` (Yu, 2024) to do functional gene set enrichment on the gene ontology with multiple separate gene set inputs and `goenrich_heatmap` plots these results in a heatmap. Notice that there are no significant enrichments from a2, a3, and a5, so they don't appear in the heatmap. This will take a couple of minutes to run.

```{r, fig.width=8, fig.height=10}
genes <- list(a1 = TWASviz::predixcan_twas1[, "gene"],
        a2 = TWASviz::predixcan_twas2[predixcan_twas1$pvalue < 0.05, "gene"],
        a3 = TWASviz::predixcan_twas3[predixcan_twas1$pvalue < 0.05, "gene"],
        a4 = TWASviz::predixcan_twas4[predixcan_twas1$pvalue < 0.001, "gene"],
        a5 = TWASviz::predixcan_twas5[predixcan_twas1$pvalue < 0.05, "gene"])
enrich_res <- TWASviz::gene_enrichment(genes)

TWASviz::goenrich_heatmap(enrich_res = enrich_res)
```

The entirety of this Application 1 can be done in the R shiny app. Just run *run_TWAS_analysis()* and follow the instructions.

## Application 2

Let's look at an analysis that performs TWAS with this package, starting from `PrediXcan` predicted gene expression.

### Visualizing your Predicted Gene Expression Data with a Heatmap
`PrediXcan` predicted gene expression output will likely give you expression levels for hundreds or thousands of genes. A nice way to look at the correlation structure of your predicted gene expressions is with a correlation heatmap which shows the pairwise correlation of each gene's expression levels. However, laptop screens are often only 1920 x 1080 pixels. How can I make a correlation matrix with more cells than there are pixels on my screen? Often, your plot viewer will do some image down-sampling to make your detailed heatmap to fit into fewer pixels. With `corr_heatmap`, you can specify the function with which your heatmap is down-sampled through the summary_func argument.
Compare these two heatmaps:

```{r, fig.width=3}
# use simulated data since the provided PrediXcan predicted gene expression data only has 100 columns per cellX_gene_expr object in order to make the below TWAS examples run faster.

big_gene_expr <- matrix(rgamma(100^2, 2), nrow = 100) %*% matrix(rnorm(200^2), ncol = 400)

# only use 1 core here because vignette builder does not like a parallel backend
corr_heatmap(gene_expr = big_gene_expr, summary_func = mean, 
             title = "Summarized with mean", n_core = 1)

corr_heatmap(gene_expr = big_gene_expr, 
             summary_func =  function(x) x[which.max(abs(x))], 
             title = "Summarized with maximum magnitude", n_core = 1)

```

The heatmap summarized with the mean is possibly more faithful to the actual magnitude of the correlation patterns in the data, but the heatmap summarized with the maximum absolute value highlights weak correlation structures that may still be important.

### Formatting PrediXcan to Perform Sparse Group Lasso TWAS
The sparse group lasso function needs a specially formatted predictor matrix and grouping information. `preprocess_expressions_pathways` duplicates genes in your imputed gene expression matrix that appear multiple times in your specified pathways, renames genes to gene_pathway, subsets and reorders your gene expression and phenotype files to match the participant ID order, and regresses out the covariates from your phenotype if you specify family_func = "gaussian" or provides offsets from predicted values from glm(phenotype ~ ., data = covars, family = family_func).

```{r}
set.seed(410)
# for one predicted gene expression dataset
# sgl_input1 <- TWASviz::preprocess_expressions_pathways(TWASviz::cell1_gene_expr, all_pathways = TWASviz::all_pathways,
# phenotype_filename = "../inst/extdata/phenotype_covars.txt", phenotype_colname = "y_bin", family_func = "binomial")

# for multiple tissue/cell type specific gene expressions
all_gene_expr <- list(TWASviz::cell1_gene_expr, TWASviz::cell2_gene_expr, TWASviz::cell3_gene_expr, TWASviz::cell4_gene_expr, TWASviz::cell5_gene_expr, TWASviz::cell6_gene_expr)
all_sgl_input <- lapply(all_gene_expr, TWASviz::preprocess_expressions_pathways, all_pathways = TWASviz::all_pathways,
phenotype_filename = "../inst/extdata/phenotype_covars.txt", phenotype_colname = "y_bin", family_func = "binomial")

```

### Performing Sparse Group Lasso TWAS
In contrast with the TWAS method available in PrediXcan (which estimates each gene expression effect size one by one with just a simple regression using the phenotype as the response and the gene expression as the predictor), using sparse group lasso is a pathway aware method of estimating the effect sizes of genes and pathways on your phenotype of interest. Briefly, sparse group lasso selects features (gene expressions) and regularizes their estimated effect sizes within groupings (pathways), and with respect to the entire feature space. 

This example shows how to use `sgl_TWAS` for a binary phenotype of interest, which is consistent with the above example that shows how to use `preprocess_expressions_pathways` (this will take a few minutes to run).
```{r}
set.seed(410)
# for one predicted gene expression dataset
# sgl_fit1 <- TWASviz::sgl_TWAS(X = sgl_input1$X, y = sgl_input1$y, 
# grouping = sgl_input1$groups, family_func = "binomial", 
# offsets = sgl_input1$offsets, pred_loss = "misclass")

# for multiple tissue/cell type specific gene expressions
sgl_fits <- lapply(all_sgl_input, function(sgl_input){
  TWASviz::sgl_TWAS(X = sgl_input$X, y = sgl_input$y, grouping = sgl_input$groups, family_func = "binomial", offsets = sgl_input$offsets, 
                              pred_loss = "misclass")})
```

### Reformat sgl_TWAS Output
`sgl_TWAS` will give you the fit sparse group lasso object and the coefficient estimates at the *lambda.1se* value (the most regularized model within one standard error of the minimum CV error). This is the suggested set of coefficients to use, but you can specify other coefficient sets that are more or less regularized. 

The coefficient format from `sgl_TWAS` is not very useful. To make these results compatible with other functions in TWASviz, use `sgl2adj_df` to reformat your many TWAS associations into an adjacency matrix. Or, you can use `sgl2txt_file` to make your sparse group lasso output a text file that is compatible with the R shiny app (*run_TWAS_analysis()*). The *NA* in the `adj_dfs` output means that the sparse group lasso did not select this gene/pathway in this as strongly associated with your phenotype of interest, but it is associated in a different tissue.

```{r, warning=FALSE, message=FALSE}
# use your actual cell type or tissue names
tissue_names <- LETTERS[seq_along(sgl_fits)]

# combine all the lambda1se_coef from sgl_fits
all_coefs <- lapply(sgl_fits, function(x){
  x$lambda1se_coef
})
names(all_coefs) <- tissue_names

# for use in correlation_overlap_heatmap
adj_dfs <- TWASviz::sgl2adj_df(all_coefs)

head(adj_dfs$gene)

head(adj_dfs$pathway)


# for usability outside of R and in the shiny app
# commented out because it writes files
# sgl_df <- mapply(TWASviz::sgl2txt_file, 
#                  all_coefs, paste0("../inst/extdata/", 
#                                    tissue_names, "_sgl_TWAS_"))
```

Some of the cell type specific predicted gene expressions don't have any genes selected from the sparse group lasso association with the phenotype, after adjusting for covariates. This is meaningful, as this provides evidence for the importance/unimportance of each cell type in the transcriptomic basis of your phenotype of interest.

Then you can use `adj_dfs$gene` and `adj_dfs$pathway` in the analysis outlined in Approach 1 to get nice looking gene level and pathway level overlap plots, respectively, and GO enrichment plots from `adj_dfs$gene`.


## References

Aleksander, S. A., Balhoff, J., Carbon, S., Cherry, J. M., Drabkin, H. J., Ebert, D., Feuermann, M., Gaudet, P., Harris, N. L., Hill, D. P., Lee, R., Mi, H., Moxon, S., Mungall, C. J., Muruganugan, A., Mushayahama, T., Sternberg, P. W., Thomas, P. D., Van Auken, K., … Westerfield, M. (2023). The gene ontology knowledgebase in 2023. GENETICS, 224(1). https://doi.org/10.1093/genetics/iyad031 

Ashburner, M., Ball, C. A., Blake, J. A., Botstein, D., Butler, H., Cherry, J. M., Davis, A. P., Dolinski, K., Dwight, S. S., Eppig, J. T., Harris, M. A., Hill, D. P., Issel-Tarver, L., Kasarskis, A., Lewis, S., Matese, J. C., Richardson, J. E., Ringwald, M., Rubin, G. M., & Sherlock, G. (2000). Gene ontology: Tool for the unification of biology. Nature Genetics, 25(1), 25–29. https://doi.org/10.1038/75556 

Bache, S., Wickham, H. (2022). magrittr: A Forward-Pipe Operator for R. R package version 2.0.3. https://CRAN.R-project.org/package=magrittr.

Barrett, T., Dowle, M., Srinivasan, A., Gorecki, J., Chirico, M., Hocking, T., Schwendinger, B., Krylov, I. (2025). data.table: Extension of `data.frame`. R package version 1.17.8, https://CRAN.R-project.org/package=data.table.

Bates, D., Maechler, M., Jagan, M. (2025). Matrix: Sparse and Dense Matrix Classes and Methods. R package version 1.7-4. https://CRAN.R-project.org/package=Matrix.

BioRender.com. BioRender. https://www.biorender.com (accessed 3 November 2025)
          
Campitelli, E. (2025). ggnewscale: Multiple Fill and Colour Scales in 'ggplot2'. R package version 0.5.2. https://CRAN.R-project.org/package=ggnewscale.

Carlson M (2024). org.Hs.eg.db: Genome wide annotation for Human. R package version 3.20.0.

Chang, W., Cheng, J., Allaire, J., Sievert, C., Schloerke, B., Xie, Y., Allen, J., McPherson, J., Dipert, A., Borges, B. (2025). shiny: Web Application Framework for R. R package version 1.11.1. https://CRAN.R-project.org/package=shiny.

Devailly, G.  (2021, October 14). Plotting heatmaps from big matrices in R. https://gdevailly.netlify.app/post/plotting-big-matrices-in-r/

Diamant, I., Clarke, D. J. B., Evangelista, J. E., Lingam, N., & Ma’ayan, A. (2024). Harmonizome 3.0: Integrated knowledge about genes and proteins from diverse multi-omics resources. Nucleic Acids Research, 53(D1). https://doi.org/10.1093/nar/gkae1080 

Gamazon, E. R., Wheeler, H. E., Shah, K. P., Mozaffari, S. V., Aquino-Michaels, K., Carroll, R. J., Eyler, A. E., Denny, J. C., Nicolae, D. L., Cox, N. J., & Im, H. K. (2015). A gene-based association method for mapping traits using reference transcriptome data. Nature Genetics, 47(9), 1091–1098. https://doi.org/10.1038/ng.3367 

Kanehisa, M. (2000). Kegg: Kyoto encyclopedia of genes and genomes. Nucleic Acids Research, 28(1), 27–30. https://doi.org/10.1093/nar/28.1.27 

Liang, X., Cohen, A., Solón Heinsfeld, A., Pestilli, F., McDonald, D.J. (2024). “sparsegl: An R Package for Estimating Sparse Group Lasso.” Journal of Statistical Software, 110(6), 1-23. doi:10.18637/jss.v110.i06. https://doi.org/10.18637/jss.v110.i06.

Milacic, M., Beavers, D., Conley, P., Gong, C., Gillespie, M., Griss, J., Haw, R., Jassal, B., Matthews, L., May, B., Petryszak, R., Ragueneau, E., Rothfels, K., Sevilla, C., Shamovsky, V., Stephan, R., Tiwari, K., Varusai, T., Weiser, J., … D’Eustachio, P. (2023). The reactome pathway knowledgebase 2024. Nucleic Acids Research, 52(D1). https://doi.org/10.1093/nar/gkad1025 

OpenAI. (2025). ChatGPT (GPT-5) large language model. https://chat.openai.com/ (accessed 28 November 2025)

R Core Team (2024). R: A Language and Environment for Statistical Computing. R Foundation for Statistical Computing, Vienna, Austria. https://www.R-project.org/.
  
Revelle, W. (2025). psych: Procedures for Psychological, Psychometric, and Personality Research. Northwestern University, Evanston, Illinois. R package version 2.5.6, https://CRAN.R-project.org/package=psych.

Simon, N., Friedman, J., Hastie, T., & Tibshirani, R. (2013). A sparse-group lasso. Journal of Computational and Graphical Statistics, 22(2), 231–245. https://doi.org/10.1080/10618600.2012.681250 

Slowikowski K (2024). ggrepel: Automatically Position Non-Overlapping Text Labels with 'ggplot2'. R package version 0.9.6. https://CRAN.R-project.org/package=ggrepel.

Storey, J. D., & Tibshirani, R. (2003). Statistical significance for genomewide studies. Proceedings of the National Academy of Sciences, 100(16), 9440–9445. https://doi.org/10.1073/pnas.1530509100 

Tabsets. Shiny. (2014, July 30). https://shiny.posit.co/r/gallery/application-layout/tabsets 

Wainberg, M., Sinnott-Armstrong, N., Mancuso, N., Barbeira, A. N., Knowles, D. A., Golan, D., Ermel, R., Ruusalepp, A., Quertermous, T., Hao, K., Björkegren, J. L., Im, H. K., Pasaniuc, B., Rivas, M. A., & Kundaje, A. (2019). Opportunities and challenges for transcriptome-wide association studies. Nature Genetics, 51(4), 592–599. https://doi.org/10.1038/s41588-019-0385-z 

Wang, N., Ye, Z., & Ma, T. (2024). Tips: A novel pathway-guided joint model for transcriptome-wide association studies. Briefings in Bioinformatics, 25(6). https://doi.org/10.1093/bib/bbae587 

Wickham, H., François, R., Henry, L., Müller, K., Vaughan, D. (2023). dplyr: A Grammar of Data Manipulation. R package version 1.1.4. https://CRAN.R-project.org/package=dplyr.

Wickham, H. ggplot2: Elegant Graphics for Data Analysis. Springer-Verlag New York, 2016.

Wickham, H. (2025). stringr: Simple, Consistent Wrappers for Common String Operations. R package version 1.6.0. https://CRAN.R-project.org/package=stringr.

Wickham, H., Vaughan, D., Girlich, M. (2024). tidyr: Tidy Messy Data. R package version 1.3.1. https://CRAN.R-project.org/package=tidyr.

Xu, S., Hu, E., Cai, Y., Xie, Z., Luo, X., Zhan, L., Tang, W., Wang, Q., Liu, B., Wang, R., Xie, W., Wu, T., Xie, L., Yu, G. Using clusterProfiler to characterize multiomics data. Nature Protocols. 2024, 19(11):3292-3320.
------------------------------------------------------------------------

```{r}
sessionInfo()
```
