---
title: "Introduction to TWASviz"
author: "Megan Ding"
date: "2025-11-28"
output:
  rmarkdown::html_vignette:
    toc: true
    number_sections: false
vignette: >
  %\VignetteIndexEntry{Introduction_TWASviz}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
## Introduction
`TWASviz` is an R package designed to produce important data exploration plots and Transcriptome Wise Association Studies (TWASs) with sparse group lasso (Simon et al., 2013). TWASs use imputed gene expression data and test for association between each gene’s expression levels against phenotype data. This process allows us to identify transcripts that are significantly correlated to a phenotype of interest. TWASviz will allow users to perform TWASs using sparse group lasso in R (Wang et al., 2024), and import their own TWAS results, and effectively visualize the output. 

This package is designed to use perform sparse group lasso TWASs on individual imputed gene expression data from the external software `PrediXcan` (Gamazon et al., 2015), as well as to process the linear regression TWAS output from `PrediXcan`. For more information on `PrediXcan` and its individual level gene expression imputation and TWAS association tests visit https://github.com/hakyimlab/MetaXcan/wiki/Individual-level-PrediXcan:-introduction,-tutorials-and-manual .

To download **TWASviz**, use the following commands:

```{r}
if (!requireNamespace("devtools")){
  install.packages("devtools", repos = "http://cran.us.r-project.org")
}
require("devtools")
devtools::install_github("mding010705/TWASviz", build_vignettes = TRUE)
library("TWASviz")
```

To list all functions available in the package:
```{r}
# ls("package:TWASviz") Lists all the functions from imported packages as well (too many)

# to list the exported functions in TWASviz
getNamespaceExports("TWASviz")
```

To list all sample datasets available in the package:
```{r}
data(package = "TWASviz")
```


## Components

![TWASviz workflow.](../inst/extdata/TWASviz_workflow.png)

## Application 1

Let's look at an analysis that starts from `PrediXcan` TWAS association output.

### Loading PrediXcan Association Results

The package comes with sample prediXcan association results (`predixcan_twasX` for sample `PrediXcan` association output).

```{r}
head(TWASviz::predixcan_twas1)
head(TWASviz::predixcan_twas2)
head(TWASviz::predixcan_twas3)
```

### Formatting Multiple PrediXcan TWAS Results

`predixcan2adj_df` is a function that reads in multiple `PrediXcan` association files and formats them in an adjacency matrix like dataframe, thresholding on the gene associations' p-values.


```{r}
adj_df <- TWASviz::predixcan2adj_df(predixcan_assoc_filenames = paste0("../inst/extdata/predixcan_twas", 1:6, ".txt"), 
                           pvalue_thresh = 0.02)

head(adj_df)
```


### Understanding the Relationship between Multiple TWASs

The `correlation_overlap_heatmap` function provides a heatmap for pairwise complete correlations for effect sizes of multiple gene sets (from TWASs) in the lower triangle, and number of pairwise complete effect sizes for multiple gene sets in the upper triangle. The diagonal displays the total number of effect sizes for a single gene set.

```{r}
TWASviz::correlation_overlap_heatmap(adj_df, tissue_names = c("a", "b", "c",
                                                     "d", "e", "f"))

```

### Understanding Significance and Effect Size Relationship 

`volcano_plot` produces a volcano plot of the effect size z-score against the -log10(pvalue). It labels and colors significant genes. The horizontal line is the significance threshold, and the vertical lines are at the normal 2.5 and 97.5th quantile thresholds. Notice that with `PrediXcan` association outputs, the systematic relationship between effect size and p-value is very noticeable due to the constant effect size.

```{r}
TWASviz::volcano_plot(betas_p_vals = TWASviz::predixcan_twas1[1:1000, c("zscore", "pvalue", "gene")], p_thresh = 0.0006)
```

### Gene Enrichment

Having a set of significant genes is not useful when you can't summarize their functional associations. `gene_enrichment` allows you to use `clusterProfiler` (Yu, 2024) to do functional gene set enrichment on the gene ontology with multiple separate gene set inputs and `goenrich_heatmap` plots these results in a heatmap. Notice that there are no significant enrichments from a2, a3, and a5, so they don't appear in the heatmap. This will take a couple of minutes to run.

```{r}
genes <- list(a1 = TWASviz::predixcan_twas1[, "gene"],
        a2 = TWASviz::predixcan_twas2[predixcan_twas1$pvalue < 0.05, "gene"],
        a3 = TWASviz::predixcan_twas3[predixcan_twas1$pvalue < 0.05, "gene"],
        a4 = TWASviz::predixcan_twas4[predixcan_twas1$pvalue < 0.001, "gene"],
        a5 = TWASviz::predixcan_twas5[predixcan_twas1$pvalue < 0.05, "gene"])
enrich_res <- TWASviz::gene_enrichment(genes)

TWASviz::goenrich_heatmap(enrich_res = enrich_res)
```


## Application 2

Let's look at an analysis that performs TWAS with this package, starting from `PrediXcan` predicted gene expression.

### Visualizing your Predicted Gene Expression Data with a Heatmap
`PrediXcan` predicted gene expression output will likely give you expression levels for hundreds or thousands of genes. A nice way to look at the correlation structure of your predicted gene expressions is with a correlation heatmap which shows the pairwise correlation of each gene's expression levels. However, laptop screens are often only 1920 x 1080 pixels. How can I make a correlation matrix with more cells than there are pixels on my screen? Often, your plot viewer will do some image down-sampling to make your detailed heatmap to fit into fewer pixels. With `corr_heatmap`, you can specify the function with which your heatmap is down-sampled through the summary_func argument.
Compare these two heatmaps (this will take a few minutes to run):

```{r}
# use simulated data since the provided PrediXcan predicted gene expression data only has 100 columns per cellX_gene_expr object in order to make the below TWAS examples run faster.

big_gene_expr <- matrix(rgamma(100^2, 2), nrow = 100) %*% matrix(rnorm(200^2), ncol = 400)
corr_heatmap(gene_expr = big_gene_expr, summary_func = mean, 
             title = "Summarized with mean")

corr_heatmap(gene_expr = big_gene_expr, 
             summary_func =  function(x) x[which.max(abs(x))], 
             title = "Summarized with maximum magnitude")
```
The heatmap summarized with the mean is possibly more faithful to the actual magnitude of the correlation patterns in the data, but the heatmap summarized with the maximum absolute value highlights weak correlation structures that may still be important.

### Formatting PrediXcan to Perform Sparse Group Lasso TWAS
The sparse group lasso function needs a specially formatted predictor matrix and grouping information. `preprocess_expressions_pathways` duplicates genes in your imputed gene expression matrix that appear multiple times in your specified pathways, renames genes to gene_pathway, subsets and reorders your gene expression and phenotype files to match the participant ID order, and regresses out the covariates from your phenotype if you specify family_func = "gaussian" or provides offsets from predicted values from glm(phenotype ~ ., data = covars, family = family_func).

```{r}
set.seed(410)
# for one predicted gene expression dataset
sgl_input1 <- TWASviz::preprocess_expressions_pathways(TWASviz::cell1_gene_expr, all_pathways = TWASviz::all_pathways,
phenotype_filename = "../inst/extdata/phenotype_covars.txt", phenotype_colname = "y_bin", family_func = "binomial")

# for multiple tissue/cell type specific gene expressions
all_gene_expr <- list(TWASviz::cell1_gene_expr, TWASviz::cell2_gene_expr, TWASviz::cell3_gene_expr, TWASviz::cell4_gene_expr, TWASviz::cell5_gene_expr, TWASviz::cell6_gene_expr)
all_sgl_input <- lapply(all_gene_expr, TWASviz::preprocess_expressions_pathways, all_pathways = TWASviz::all_pathways,
phenotype_filename = "../inst/extdata/phenotype_covars.txt", phenotype_colname = "y_bin", family_func = "binomial")

```

### Performing Sparse Group Lasso TWAS
In contrast with the TWAS method available in PrediXcan (which estimates each gene expression effect size one by one with just a simple regression using the phenotype as the response and the gene expression as the predictor), using sparse group lasso is a pathway aware method of estimating the effect sizes of genes and pathways on your phenotype of interest. Briefly, sparse group lasso selects features (gene expressions) and regularizes their estimated effect sizes within groupings (pathways), and with respect to the entire feature space. 

This example shows how to use `sgl_TWAS` for a binary phenotype of interest, which is consistent with the above example that shows how to use `preprocess_expressions_pathways` (this will take a few minutes to run).
```{r}
set.seed(410)
# for one predicted gene expression dataset
sgl_fit1 <- TWASviz::sgl_TWAS(X = sgl_input1$X, y = sgl_input1$y, grouping = sgl_input1$groups, family_func = "binomial", offsets = sgl_input1$offsets, 
                              pred_loss = "misclass")

# for multiple tissue/cell type specific gene expressions
sgl_fits <- lapply(all_sgl_input, function(sgl_input){
  TWASviz::sgl_TWAS(X = sgl_input$X, y = sgl_input$y, grouping = sgl_input$groups, family_func = "binomial", offsets = sgl_input$offsets, 
                              pred_loss = "misclass")})
```

### Reformat sgl_TWAS Output
`sgl_TWAS` will give you the fit sparse group lasso object and the coefficient estimates at the *lambda.1se* value (the most regularized model within one standard error of the minimum CV error). This is the suggested set of coefficients to use, but you can specify other coefficient sets that are more or less regularized. 

The coefficient format from `sgl_TWAS` is not very useful. To make these results compatible with other functions in TWASviz, use `sgl2adj_df` to reformat your many TWAS associations into an adjacency matrix.

```{r}
all_coefs <- lapply(sgl_fits, function(x){
  x$lambda1se_coef
})
adj_dfs <- TWASviz::sgl2adj_df(all_coefs)

head(adj_dfs)
```

Some of the cell type specific predicted gene expressions don't have any genes selected from the sparse group lasso association with the phenotype, after adjusting for covariates. This is meaningful, as this provides evidence for the importance/unimportance of each cell type in the transcriptomic basis of your phenotype of interest.

Then you can use `adj_dfs$gene` and `adj_dfs$pathway` in the analysis outlined in Approach 1 to get nice looking gene level and pathway level overlap plots, respectively, and GO enrichment plots from `adj_dfs$gene`.


## References
Gamazon, E. R., Wheeler, H. E., Shah, K. P., Mozaffari, S. V., Aquino-Michaels, K., Carroll, R. J., Eyler, A. E., Denny, J. C., Nicolae, D. L., Cox, N. J., & Im, H. K. (2015). A gene-based association method for mapping traits using reference transcriptome data. Nature Genetics, 47(9), 1091–1098. https://doi.org/10.1038/ng.3367 

Simon, N., Friedman, J., Hastie, T., & Tibshirani, R. (2013). A sparse-group lasso. Journal of Computational and Graphical Statistics, 22(2), 231–245. https://doi.org/10.1080/10618600.2012.681250 

Yu G (2024). “Thirteen years of clusterProfiler.” The Innovation, 5(6), 100722. doi:10.1016/j.xinn.2024.100722, https://doi.org/10.1016/j.xinn.2024.100722.

Wang, N., Ye, Z., & Ma, T. (2024). Tips: A novel pathway-guided joint model for transcriptome-wide association studies. Briefings in Bioinformatics, 25(6). https://doi.org/10.1093/bib/bbae587 
------------------------------------------------------------------------

```{r}
sessionInfo()
```
